Add syntax for tables

gpetiot opened the pull request:
This is a first draft, I'm asking for feedback before going further. @jonludlam @Julow

I chose this syntax:

```
{t
  {tr
     {td ...}
     ...
     {td ...}
  }
  ...
  {tr
     {td ...}
     ...
     {td ...}
  }
}
```
The markup is ovsiously inspired from the `<tr></tr><td></td>` HTML tags.
The generated output is straightforward in HTML and Latex, at first I wanted to add header cells (`<th></th>`) but I'm not aware of anything similar in Latex but we could apply a different style.

Some improvements:
- add the ability to say that a cell will span over N>1 columns. In HTML it would be `<td colspan="N">...</td>`, in Latex it would be `\multicolumn{N}{c}{...}`
- check that each row has the same number of columns? (I don't remember if that will cause issues for Latex, but HTML should be more lenient).

What do you think?

gpetiot committed:
Add syntax for tables

jonludlam commented:
Thanks @gpetiot !

This is definitely an approach that could work very well. However, I think we ought to consider the possibility of having more visual way to layout tables, perhaps something similar to how rst or orgmode does it - as an example both of the following are valid rst tables:

```
+------------+------------+-----------+
| Header 1   | Header 2   | Header 3  |
+============+============+===========+
| body row 1 | column 2   | column 3  |
+------------+------------+-----------+
| body row 2 | Cells may span columns.|
+------------+------------+-----------+
| body row 3 | Cells may  | - Cells   |
+------------+ span rows. | - contain |
| body row 4 |            | - blocks. |
+------------+------------+-----------+
```

and 

```
=====  =====  ======
   Inputs     Output
------------  ------
  A      B    A or B
=====  =====  ======
False  False  False
True   False  True
False  True   True
True   True   True
=====  =====  ======
```

and here's an org-mode table:

```
| N | N^2 | N^3 | N^4 | sqrt(n) | sqrt[4](N) |
|---+-----+-----+-----+---------+------------|
| / |  <  |     |  >  |       < |          > |
| 1 |  1  |  1  |  1  |       1 |          1 |
| 2 |  4  |  8  | 16  |  1.4142 |     1.1892 |
| 3 |  9  | 27  | 81  |  1.7321 |     1.3161 |
|---+-----+-----+-----+---------+------------|
```


The reason being that more often than not right now I'm still looking at docs in my editor buffer, or on github, as opposed to nicely rendered on a web page. We'd also get to use pre-existing table-editing modes in various editors.

dbuenzli commented:
Most lightweight markup languages add two things, headers and specification of column alignement (see for example the github flavoured markdown [table definition](https://github.github.com/gfm/#tables-extension-)). I think it would be nice to have these. I wouldn't bother too much with multi-columns.

I have mixed feelings about what @jonludlam suggests. Of course these tables look nicer but the given examples are also not very content heavy.

Once you start adding links, references and ocamldoc markup in the cell contents I have the impression that the regular, well-delimited, syntax of ocamldoc scales better.

Julow commented:
> The reason being that more often than not right now I'm still looking at docs in my editor buffer, or on github, as opposed to nicely rendered on a web page.

I've always hated both reading and writing tables drawn and aligned like that. The editor plugins to draw them are not worth learning for the few tables I ever needed to make so I end up not aligning them, they become write-only. Also, multi line cells are a pain.

jonludlam commented:
@dbuenzli agreed that larger-scale tables won't look as nice as those examples! Nothing stopping us from doing both though - we already have precedent in lists, where we can write both
```
(** Here is a {b list}
- item 1
- item 2
- item 3

The list is ended by the blank line.*)

```
and
```
(** Here is a {b list}
{ul {- item 1}
{- item 2}
{- item 3}}
The list is ended by the blank line.*)
```

we could simply choose based on the content of the `{table ...}` element - e.g.

```
{table
 {tr
  {th header 1}
  {th header 2}}
 {tr
  {td cell 1}
  {td cell 2}}}
```

vs

```
{table
+==========+==========+
| header 1 | header 2 |
+==========+==========+
| cell 1   | cell 2   |
+----------+----------+
}
```

jonludlam commented:
Actually, choosing based on contents is not a good idea. We could use a different tag though - `{t` vs `{table` for example.

dbuenzli commented:
To motivate with concrete examples here's a sample of tables that I currently write as code and that would benefit from the proposal:

1. https://erratique.ch/software/b0/doc/B00_ocaml/Mod/Name/index.html#mangled
2. https://erratique.ch/software/brr/doc/ffi_cookbook.html#names
3. https://erratique.ch/software/brr/doc/ffi_manual.html#equality
4. https://erratique.ch/software/gg/doc/Gg/Float/index.html#floatrecall
5. https://erratique.ch/software/uutf/doc/Uutf/index.html#val-decoder 
6. https://erratique.ch/repos/inane/tree/src/b36uid.mli#n12

Scroll a bit if you don't see the table, which reminds me: it would be nice to be able to reference them via a label like we do for sections, i.e. `{table:label` (and why not cells and/or rows, but let's start simple).

dbuenzli commented:
Regarding a possible lightweight notation. I think it's good at least have a look at what the [experts do](https://htmlpreview.github.io/?https://github.com/jgm/djot/blob/master/doc/syntax.html#pipe-table) to when they can start designing from scratch.

gpetiot commented:
For the light syntax, what do you think about something like:

```
{t
+-----------------------------
|{l}  xx |{c}  yy  |{r}   zz
+---------------------------
|   xx   |{r}   yy  |{l}    zz
+-----------------------------
}
```

- No need to finish the line (just `+-` at the beginning of the line to indicate a new row)
- `{l}`/`{r}`/`{c}` could indicate the alignment (optional), but should be stuck to the `|` to be a special token

dbuenzli commented:
> For the light syntax, what do you think about something like:

If you really want to add this then I strongly suggest not to invent anything new here. Have a look at [pandoc's markdown table extensions](https://pandoc.org/MANUAL.html#tables) for a few alternatives. 

Personally I think you should just go with `pipe_tables` which is popular among a few extended markdown dialects including [github's](https://github.github.com/gfm/#tables-extension-) and [gitlab's](https://docs.gitlab.com/ee/user/markdown.html#tables) ones.

gpetiot commented:
I implemented the light and heavy syntaxes, this PR is ready to be reviewed.

Julow commented on src/syntax.ml:
```diff
@@ -1067,6 +1149,354 @@ and explicit_list_items :
+      |> add_warning input;
+      ([], [], [], brace_location)
+  | header :: [] ->
+      Parse_error.not_allowed ~what:"missing separator between header and data"
```
This concatenates "is not allowed", which is not very nice. It should be a new kind of error.

gpetiot replied:
I rewrote this pattern matching a, more things are allowed now (no header, no align row)

Julow commented on src/syntax.ml:
```diff
@@ -1067,6 +1149,354 @@ and explicit_list_items :
+        brace_location
+      |> add_warning input;
+      (header, [], [], brace_location)
+  | header :: align :: data ->
```
This is arbitrary. I want tables without a header.

gpetiot replied:
done

Julow commented on src/syntax.ml:
```diff
@@ -1067,6 +1149,354 @@ and explicit_list_items :
+        let ws = List.map (fun x -> { Loc.value = `Word x; location }) ws in
+        if List.for_all (fun x -> Option.is_some (valid_align x)) ws then
+          (* revert the list for consistency *)
+          Some (List.map (fun x -> [ x ]) ws |> List.rev)
```
The extra wrapping in a list adds some "impossible" cases later. I'd prefer if the separator row had its own more comfortable type. It doesn't need to be mixed with the other rows later, the caller code needs to be changed.

Same for the reversed list, it feels wrong.

gpetiot replied:
done

Julow commented on src/syntax.ml:
```diff
@@ -1067,6 +1149,354 @@ and explicit_list_items :
+          else w
+        in
+        let ws = String.split_on_char '|' w in
+        let ws = List.map (fun x -> { Loc.value = `Word x; location }) ws in
```
`valid_align` doesn't need to work on `` `Word ``, removing a layer of match and simplifying the `List.map`s.

gpetiot replied:
done

Julow commented on src/syntax.ml:
```diff
@@ -1067,6 +1149,354 @@ and explicit_list_items :
+  in
+  let acc_r = match acc_c with [] -> acc_r | _ -> List.rev acc_c :: acc_r in
+  let acc_t = match acc_r with [] -> acc_t | _ -> List.rev acc_r :: acc_t in
+  let rows = List.rev acc_t in
```
`acc_t`, `acc_r`, `acc_c` that's too hard to follow and adds unnecessary constraints: separator row validated once then parsed again, "impossible" cases, duplicated code.

I suggest making two "consume" functions: One for the header and separator, the other for the remaining rows. That way, the parsing of the first two is not mentioned all over the place and can return the more interesting type directly.
I believe this can also help to support an optional header row, and remove the duplicated error checking.

gpetiot replied:
I've split the consume function in two. But if we would change the design to first expect the header, then the align row, then the data it would not allow to have tables containing a single line, or only data, etc. I find it convenient to have the align row have the same type as the data (represented as Word) and interpret it when we parsed everything (allow more flexible tables). The code has been heavily modified, what do you think?

Julow reviewed (commented):
This parses as two columns, `a` and `b |c| d`, which is not intuitive:

```
{t
 | a | b |c| d |
 |---|--:|:--|:-:|
}
```

jonludlam commented:
I assume that, since pipetables are pretty standard, there are a set of nasty test cases somewhere that we could use?

dbuenzli commented:
Would it please be possible to update the first comment with the implemented syntaxes. For example how do I specify alignement in the "heavy" syntax ?

gpetiot commented:
> Would it please be possible to update the first comment with the implemented syntaxes. For example how do I specify alignement in the "heavy" syntax ?

Nothing decided yet, I'm torn between doing something similar to the light syntax:
```
{table
   {tr {th ...} [th ...}}
   {align left right}
   ...
}
```
or something that would be easier to read to my opinion but that would make both implementations diverge:
```
{table
  {tr {th{left} ...} {th{right} ...}} or {tr {th:left ...} {th:right ...}}
  ...
}
```

dbuenzli commented:
Note that even though IIRC `odoc` dropped support for it. There was a syntax in `ocamldoc` for [text alignement](https://v2.ocaml.org/manual/ocamldoc.html#inline-text-element). Might be worth resurecting it for `th`. Something like: a bare `th` is left-aligned (or whathever the default is of the light syntax) and otherwise alignement is specified by "direct" nesting:

```
{tr
{th {L Left aligned}} {th {R Right aligned}} {th {C Center aligned}} {th default align}}
```

gpetiot commented:
> Note that even though IIRC `odoc` dropped support for it. There was a syntax in `ocamldoc` for [text alignement](https://v2.ocaml.org/manual/ocamldoc.html#inline-text-element). Might be worth resurecting it for `th`. Something like: a bare `th` is left-aligned (or whathever the default is of the light syntax) and otherwise alignement is specified by "direct" nesting:
> 
> ```
> {tr
> {th {L Left aligned}} {th {R Right aligned}} {th {C Center aligned}} {th default align}}
> ```

Sounds good to me, @jonludlam what do you think?

jonludlam commented:
Reusing the syntax we previously had makes sense to me.

Julow commented:
On the other hand, an attribute-like syntax would allow `colspan` and keep room for new things later.

Julow commented on src/compat.ml:
```diff
@@ -0,0 +1,40 @@
+    | Some x -> x
+
+  let join_list l =
+    if List.for_all Option.is_some l then Some (List.map value_exn l) else None
```
Using the stdlib's module, should be:
```ocaml
    if List.for_all is_some l then Some (List.map value_exn l) else None
```

Julow commented on src/syntax.ml:
```diff
@@ -1067,6 +1153,358 @@ and explicit_list_items :
+        if String.length w = 1 then
+          match w with "-" -> Some `Default | ":" -> Some `Center | _ -> None
+        else if
+          String.for_all (Char.equal '-') (String.sub w 1 (String.length w - 2))
```
This is `String.contains (...) '-'`. Two compat functions can be removed.

This function would be easier to read if it was broken into smaller functions. Also, `valid_align` could benefit from some local variables to avoid copying the string or to store the length of string.

gpetiot replied:
But I want to check that it only contains '-'.

Julow commented on src/compat.ml:
```diff
@@ -0,0 +1,40 @@
+module Char = struct
+  include Char
+
+  let equal x y = compare x y = 0
```
This is the slow polymorphic version of the function. I'd prefer if this wasn't defined. The better version is:
```ocaml
let equal (x : char) y = x = y
```

Julow commented on src/compat.ml:
```diff
@@ -0,0 +1,40 @@
+    in
+    aux 0
+
+  let split_on_char sep s =
```
It doesn't seem to be used.

Julow reviewed (commented):
I'm not a fan of the pipe syntax at all but using the most common syntax can't hurt Odoc.

gpetiot commented:
I fixed the parsing issues with the alignment tags for the heavy syntax. Let me know what you think.

panglesd commented on src/syntax.ml:
```diff
@@ -404,6 +406,8 @@ type where_in_line =
   | `After_tag
   | `After_shorthand_bullet
   | `After_explicit_list_bullet
+  | `After_table_header
+  | `After_table_cell
```
Maybe the comment above could be updated!

gpetiot replied:
done

panglesd commented on src/syntax.ml:
```diff
@@ -561,6 +569,8 @@ let rec block_element_list :
         | Top_level -> (List.rev acc, next_token, where_in_line)
         | In_shorthand_list -> (List.rev acc, next_token, where_in_line)
         | In_explicit_list -> (List.rev acc, next_token, where_in_line)
+        | In_table_header -> (List.rev acc, next_token, where_in_line)
+        | In_table_cell -> (List.rev acc, next_token, where_in_line)
```
Maybe also update the comment above the `block_element_list` function definition?

gpetiot replied:
done

Julow commented on src/ast.ml:
```diff
@@ -41,13 +52,18 @@ type nestable_block_element =
 
+and table =
+  | Light of inline_element abstract_table
+  | Heavy of nestable_block_element abstract_table
```
The ast is entirely made of structural types, this should be a polymorphic variant for consistency. Same for `abstract_table` which could be a simple tuple.

gpetiot replied:
done

Julow commented on src/syntax.ml:
```diff
@@ -1067,6 +1139,330 @@ and explicit_list_items :
+  in
+  consume_cell_items []
+
+and heavy_table_header ~parent_markup input =
```
The 3 functions have very similar code: Parsing nested-block-elements until a `` `Right_brace ``. The only change is how to handle `` `Begin_table_header | `Begin_table_data | `Begin_paragraph_style `` etc.. I think a more general higher order function is possible.

gpetiot replied:
I factorized a bit and removed some unnecessary warnings when parsing.

Julow reviewed (commented):
Nice!

Adding a block element into a light table makes quite a mess of warnings and continue the parsing in an inconsistent way:
```ocaml
# p {|
  {t
   | {[ a ]} | b |
   |---|---|
  }
  |} ;;
File "", line 2, characters 5-12:
'{[...]}' (code block) is not allowed in '{t ...}' (table).
File "", line 2, characters 13-14:
'|' is not allowed in top-level text.
Suggestion: move '|' into '{t ...}' (table).
File "", line 2, characters 15-16:
Paragraph should begin on its own line.
File "", line 2, characters 17-18:
'|' is not allowed in top-level text.
Suggestion: move '|' into '{t ...}' (table).
File "", line 3, characters 3-4:
'|' is not allowed in top-level text.
Suggestion: move '|' into '{t ...}' (table).
File "", line 3, characters 7-8:
'|' is not allowed in top-level text.
Suggestion: move '|' into '{t ...}' (table).
File "", line 3, characters 8-11:
Paragraph should begin on its own line.
File "", line 3, characters 11-12:
'|' is not allowed in top-level text.
Suggestion: move '|' into '{t ...}' (table).
File "", line 4, characters 2-3:
Unpaired '}' (end of markup).
Suggestion: try '\}'.
- : Ast.t =
[`Table (Light {header = []; data = []; align = []});
 `Paragraph [`Word "b"; `Space " "]; `Paragraph [`Word "---"];
 `Paragraph [`Word "---"]; `Paragraph [`Word "}"]]
```


The parser has the goal of not loosing text in case of bad syntax so the problematic block element shouldn't be removed. It would be nice to handle the conversion from table to paragraphs without generating too much warnings.

panglesd commented on src/syntax.ml:
```diff
@@ -36,6 +38,107 @@ let peek input =
+    consume ()
+
+  module Infix = struct
+    let ( let>> ) consume if_token =
```
this syntax is only available with ocaml 4.08. `odoc-parser` should be available starting from 4.02.

```suggestion
    let ( >>> ) consume if_token =
```

gpetiot replied:
dome

panglesd commented on src/syntax.ml:
```diff
@@ -1067,6 +1243,178 @@ and explicit_list_items :
+   which is consumed. *)
+and light_table ~parent_markup ~parent_markup_location input =
+  let rec consume_rows acc ~last_loc =
+    let>> next_token = Reader.until_rbrace input acc in
```
```suggestion
    Reader.until_rbrace input acc >>> fun next_token ->
```

panglesd commented on src/syntax.ml:
```diff
@@ -1067,6 +1243,178 @@ and explicit_list_items :
+   which is consumed. *)
+and heavy_table ~parent_markup ~parent_markup_location input =
+  let rec consume_rows acc ~last_loc =
+    let>> next_token = Reader.until_rbrace input acc in
```
```suggestion
    Reader.until_rbrace input acc >>> fun next_token ->
```

panglesd commented on src/syntax.ml:
```diff
@@ -1067,6 +1243,178 @@ and explicit_list_items :
+   which is consumed. *)
+and heavy_table_row ~parent_markup input =
+  let rec consume_cell_items acc =
+    let>> next_token = Reader.until_rbrace input acc in
```
```suggestion
    Reader.until_rbrace input acc >>> fun next_token ->
```

panglesd commented on src/syntax.ml:
```diff
@@ -1067,6 +1243,178 @@ and explicit_list_items :
+   which is consumed. *)
+and heavy_table_header ~parent_markup input =
+  let rec consume_items acc =
+    let>> next_token = Reader.until_rbrace input acc in
```
```suggestion
    Reader.until_rbrace input acc >>> fun next_token ->
```

panglesd commented on src/syntax.ml:
```diff
@@ -36,6 +38,107 @@ let peek input =
+          | 1 -> (
+              match w with
+              | "-" -> Some None
+              | ":" -> Some (Some `Center)
```
I find it strange to see an `option option`. What is the difference between `Some None` and `None`? Maybe ``Some `Unspecified`` (or a more explicit variant) should replace `Some None`?

gpetiot replied:
done

panglesd commented on src/syntax.ml:
```diff
@@ -36,6 +38,107 @@ let peek input =
+        row
+      |> Option.join_list
+
+    let from_grid grid : Ast.table =
```
The type of `from_grid` is not `_ grid -> _` (while it is for `Light_syntax.from_grid`), which is confusing...

Anyway, I think if the `cell` type is modified to include the header/normal information, are not that useful anymore.

gpetiot replied:
done

panglesd commented on test/test_tables.ml:
```diff
@@ -0,0 +1,514 @@
+          (((f.ml (1 0) (1 37))
+            (table (syntax heavy) (header ((cell ())))
+             (data ((row ((cell ()))) (row ((cell ()))))) (align (none))))))
+         (warnings ())) |}]
```
`{th ...}` fields should be respected.

gpetiot replied:
done

panglesd commented on src/ast.ml:
```diff
@@ -29,6 +30,11 @@ type inline_element =
     text. Similarly the [`Link] constructor has the link itself as first parameter
     and the second is the replacement text. *)
 
+type 'a cell = 'a with_location list
```
Cells should be able to be made header independently of the line:
```suggestion
type 'a cell = 'a with_location list * [ `Header | `Normal ]
```

gpetiot replied:
done

panglesd commented on src/syntax.ml:
```diff
@@ -381,6 +484,8 @@ let paragraph : input -> Ast.nestable_block_element with_location =
       [-], has been read, and only whitespace has been read since.
     - [`After_explicit_list_bullet], when a valid explicit bullet, such as [{li],
       has been read, and only whitespace has been read since.
+    - [`After_table_header], when a table header opening markup ('{th') has been read.
+    - [`After_table_cell], when a table cell opening markup ('{td') has been read.
```
10 lines above: "The five possible locations" -> "The seven possible locations"

gpetiot replied:
done

panglesd commented on src/syntax.ml:
```diff
@@ -594,6 +708,46 @@ let rec block_element_list :
+        junk input;
+        consume_block_elements ~parsed_a_tag where_in_line acc
+    (* Bars can never appear directly in block content.
+       They can only appear inside [{t ...}]. *)
```
It would be nice to be able to write `|` outside of `{t`. Otherwise, that is one more character that we have to escape...

That should be solved in the lexer I guess.

gpetiot replied:
done in [bd19b75](https://github.com/ocaml-doc/odoc-parser/pull/11/commits/bd19b75bd402e199bd9aee1ff4d5b7923cfa18a9)

panglesd commented on src/compat.ml:
```diff
@@ -0,0 +1,30 @@
+    | Some x -> x
+
+  let join_list l =
+    if List.for_all is_some l then Some (List.map value_exn l) else None
```
I'm not that keen on using exceptions... Why not:

```suggestion
  let join_list l =
    let rec loop acc = function
      | [] -> Some (List.rev acc)
      | Some a :: q -> loop (a :: acc) q
      | None :: _ -> None
    in
    loop [] l
```

gpetiot replied:
done

panglesd commented on src/ast.ml:
```diff
@@ -29,6 +30,11 @@ type inline_element =
+type 'a cell = 'a with_location list
+type 'a row = 'a cell list
+type 'a grid = 'a row list
+type 'a abstract_table = 'a row * 'a grid * alignment option list
```
It should be possible for cells to be header independently of the line:
```suggestion
type 'a cell = 'a with_location list * [ `Header | `Normal ]
type 'a row = 'a cell list
type 'a grid = 'a row list
type 'a abstract_table =  'a grid * alignment option list
```

gpetiot replied:
done

panglesd commented on src/ast.ml:
```diff
@@ -41,13 +47,18 @@ type nestable_block_element =
 
+and table =
+  [ `Light of inline_element abstract_table
+  | `Heavy of nestable_block_element abstract_table ]
```
I think the information of how the table was given is often irrelevant, so it should be easy to ignore! 
```suggestion
and table = nestable_block_element abstract_table * [ `Light | `Heavy ]
```

gpetiot replied:
done

panglesd reviewed (changes requested):
Nice !

I do have change requested for some things:

- I think that the heavy syntax should allow us to define any cell to be a header cell (just like in html). This allows for several use cases, with two common one being: "vertical" tables (tables where the header is a column) and tables with two "axis" (both a header line and a header column).
Fortunately, I think this will actually simplify the code a lot.
- I believe the lexer should do more work in the case of the light table. Currently, as mentioned by @Julow, opening a forbidden block inside a light table will generate a lot of warnings (where one would have sufficed) and the next closing will be wrongly paired. That's a major issue!
Using a specific lexer for the inside of light tables (which is what is done for verbatim for example) also removes the constraint added by this PR to escape `|` characters outside of tables.
- I don't think the AST type should make such a clear distinction between light and heavy syntax. Similarly to `` `List`, a variant on the side could distinguish between them, the information is almost always irrelevant for consumers of the parser, it has to be easy to ignore.

Another thing: I am not a huge fan of the alignment syntax for heavy tables.
To me, it seems that `{th {R text}}` will only have an effect on the cell, by the nesting. (The fact that it has an effect on the whole columns can only be understood with the light syntax in mind).
Moreover, it prevents possible future uses of `{L }` to specify the alignment on single cell, especially on header cell (if we want that, not sure...).
Instead, we could take inspiration from latex's table (or the `col` html element), where the fact that it acts on columns is evident:
```
{table {col L C R}
  {tr
    {th left-aligned} {th centered} {th right-aligned}}
  }
}
``` 

This leaves room for deciding the line borders in odoc (but maybe that's not something we want):

```
{table {col | L C | R}
  {hline}
  {tr
    {th left-aligned} {th centered} {th right-aligned}}
  }
  {hline}
}
```

gpetiot committed:
Replace let>> with >>>

gpetiot committed:
Simplify alignment

gpetiot committed:
join_list without exception

panglesd committed:
Warn new lines in markup nested in light tables

New lines inside markup (such as `{e ...}`) were not detected, although it
breaks the light table format.

Add a warning for such cases.

Signed-off-by: Paul-Elliot <peada@free.fr>

gpetiot committed:
Convert light table contents to nestable block elements

panglesd committed:
Add suggestion on wrong content inside `{table }` and `{`tr }`

Signed-off-by: Paul-Elliot <peada@free.fr>

panglesd committed:
allow `|` char as a word outside of light table

Signed-off-by: Paul-Elliot <peada@free.fr>

panglesd committed:
adding a test for varying number of cells

Signed-off-by: Paul-Elliot <peada@free.fr>

gpetiot committed:
Do not allow alignment in heavy tables

gpetiot committed:
Typo

gpetiot committed:
Preserve td/th in the datatype

gpetiot committed:
Renaming

gpetiot committed:
Merge pull request #1 from panglesd/tables

contribution to the table PR

jonludlam commented:
I had a quick look to see whether there were any differences running over the set of libraries in odoc's driver with the `ODOC_BENCHMARK` var set to true. I found a two differences - here:

- https://github.com/ocaml/ocaml/blob/trunk/typing/typedtree.mli#L214-L219
- https://github.com/ocaml/ocaml/blob/trunk/middle_end/flambda/unbox_specialised_args.mli#L26-L31

In both instances they are actually code snippets without being enclosed in any sort of verbatim/code tags, but I was slightly surprised nevertheless. It's not a show stopper but I'd be quite interested to know why the differences are there.

panglesd committed:
Fix new paragraph starting on `|`

Signed-off-by: Paul-Elliot <peada@free.fr>

gpetiot committed:
Merge pull request #2 from panglesd/tables

Fix new paragraph starting on `|`

panglesd committed:
Fix default alignment

This commit fixes several alignment issues.

1. No alignment in a cell was encoded as an alignment to `Center.

"Centering" was certainly a typo, but there is also the issue that "no specified
alignment" should really be different than "specified to ..."

For instance (assuming `{R ...}` exists):
```
{R
  {t
    |aaa|bbb|
    |---|---|
    |xxx|yyy|}}
```
would not be right-aligned inside the table, with the previous behaviour.

This commit allows for "unspecified alignment" using an option.

2. No alignment in a table was encoded as an empty list of alignment.

This was wrong, since at some point we might want to raise warnings in case the
number of alignment and the number of columns differ.

Signed-off-by: Paul-Elliot <peada@free.fr>

gpetiot committed:
Merge pull request #3 from panglesd/tables

Fix default alignment

panglesd commented:
Thanks @jonludlam for spotting a bug! It has been fixed in 97fd2a3.

I've updated https://github.com/ocaml/odoc/pull/893 to be in sync with this PR, in case you want to review them as a whole.

gpetiot committed:
Handle '+' to separate columns in 'align' row

gpetiot committed:
Renaming data->grid in tests, data is a kind of a cell

gpetiot committed:
Revert 1199bd4

jonludlam commented on src/syntax.ml:
```diff
@@ -209,9 +324,11 @@ and delimited_inline_element_list :
     parent_markup:[< Token.t ] ->
     parent_markup_location:Loc.span ->
     requires_leading_whitespace:bool ->
+    context:inline_context ->
```
This bit of threaded state is annoying. It's also very similar to the `requires_leading_whitespace` state - I wonder if the two could be combined into a single `Context.t` or similar? I suppose the new bit is relevant to both `delimited_inline_element_list` _and_ `inline_element` so maybe not worth it? It's also possibly worth considering getting rid of the context and just checking in the `light_table` code after we've called this function that the start and end lines are the same?

jonludlam replied:
I had a quick go: https://github.com/jonludlam/odoc-parser/commit/5c49f986e1094617d566a7a1a20b0106ff954ff3 - I think it's a bit less invasive and keeps the check within the `light_table_row` function. What do you think?

gpetiot replied:
It's also my opinion that having the check in `light_table_row` makes the code clearer.

panglesd replied:
As a small remark, the location warning in the current PR and https://github.com/jonludlam/odoc-parser/commit/5c49f986e1094617d566a7a1a20b0106ff954ff3 will be different:
- In the current PR, it is only the newline. There is one warning per newline.
- In https://github.com/jonludlam/odoc-parser/commit/5c49f986e1094617d566a7a1a20b0106ff954ff3 there is only one warning, with location being the whole inline item.

I think both solutions have different advantages/disadvantages (one is clearer, the other more general), and I don't have a preference. So let's go with @jonludlam's idea!

jonludlam committed:
Alternate method of invalid light table syntax
